### ISSUE TITLE
2.10.0 Doesn't Send Events with Identical Initialization?

### ISSUE THREAD

**Issue by ScottyMJacobson** [snowplow/snowplow-javascript-tracker#704]
Hi all,

I'm looking forward to trying out the new features of 2.10 (especially useBeacon!)

However, I was hoping I'd be able to just swap out  `2.10.0` for `2.9.2`, and get up and running. For some reason, I can't get `2.10.0` to start sending events.

I uploaded these two examples, which just takes the snippet from the Setup Guide (and point at a dummy endpoint), and the only thing I change is the `2.10.0` link for the `2.9.2` one

https://snowplow2100test.netlify.com/

(in the 2.9.2 version, you can see the attempt to GET from the endpoint in the Network panel; in the 2.10.0 version, I don't see anything)

There must be something I'm missing with initializing 2.10.0, right? Sorry if it's something obvious

Thanks,
Scotty

**Comment by mhadam** [comment]
Hi Scotty,

So by default `2.10.0` will detect and fallback for the event sending method like this: Beacon (POST) -> XHR (POST) -> GET.

Opening the page you linked I noticed it triggers my ad-blocker and shutting it off allowed the tracker to send the event using Beacon.

Using POST sometimes triggers ad-blockers too, because the events are sent to a collector at the standard endpoint `/com.snowplowanalytics.snowplow/tp2`.

Let me know if this seems to be right.

Best,
Mike

**Comment by ScottyMJacobson** [comment]
Thanks Mike,

I can confirm that **in Firefox**, I'm able to see the 2.10 example sending POST network requests via sendBeacon

In Chrome, if I force the `eventMethod: 'GET'` setting, I now see 2.10 sending events via GET. [i.e. this example works on my end.](https://snowplow2100test.netlify.com/2_10_0_get.html)

However, a colleague and I tried disabling all our ad-blocking extensions **in Chrome**, on Mac and Windows, and version 71 and 72 and we still didn't see any Network activity from the 2.10 tracker via `sendBeacon`. I updated [the 2_10_0 example](https://snowplow2100test.netlify.com/2_10_0.html) to explicitly set `eventMethod: 'beacon`, and am still not seeing any activity

_(also, unrelated, but it looks like some of the documentation is out of sync. [here, it says to use beacon:true](https://github.com/snowplow/snowplow/wiki/1-General-parameters-for-the-Javascript-tracker#beacon-support), but I remember more updated documentation saying to configure `eventMethod: beacon` in the argmap (but I can't remember where) (but [here](https://github.com/snowplow/snowplow-javascript-tracker/blob/master/src/js/out_queue.js#L71) it is in the code)_

Thanks again for the assistance
Scotty

**Comment by mhadam** [comment]
Quick response while I have the chance, but I'll be looking into this more tomorrow:
It seems like Chrome preflights Beacon requests for CORS policy, so content-type: application/json requests are possibly being blocked in Chrome (whereas Firefox, et al. I'm guessing don't preflight Beacon requests).

**Comment by ScottyMJacobson** [comment]
Thanks Mike,

In light of this, we were actually in the process of forking the lib to add beacon support. Like you discovered (not sure which of these links you've seen), because of [the steps the Chromium team took](https://bugs.chromium.org/p/chromium/issues/detail?id=490015) to close [this security issue](https://github.com/dxa4481/CORS) re: preflight-less `"application/json"` beacon POSTs, when you try to send the blob with a `"application/json"` `content-type`, I think Chrome actually throws an exception (do you catch it? / is that why there's nothing in the console?)

In our fork, our solution was to "lie" in the beacon request, and send it as "plain/text", thus making it qualify as a "simple POST request" (and not require preflight); our scala collector rejected these requests, so we actually put an API Gateway between our endpoint and our collector that transformed the "content-type" header back to `"application/json"`

**Comment by mhadam** [comment]
Yes so we started catching errors in `2.10.0`. I think you can just fix this by configuring the collector to respond to the pre-flight though.

Here's an example configuration:
https://raw.githubusercontent.com/snowplow/snowplow/master/2-collectors/scala-stream-collector/examples/config.hocon.sample

Specifically:
```
crossDomain {
    enabled = false
    # Domains that are granted access, *.acme.com will match http://acme.com and http://sub.acme.com
    domains = [ "*" ]
    # Whether to only grant access to HTTPS or both HTTPS and HTTP sources
    secure = true
  }
```

**Comment by ScottyMJacobson** [comment]
We could be misunderstanding each other, but the [chromium bug thread I linked to](https://bugs.chromium.org/p/chromium/issues/detail?id=490015) (there's a lot there and it's hard to follow) seems to imply that the chromium team decided to just stonewall any attempt to send a non-simple POST request `beacon` because that would (as per the standard) necessitate a preflight request, and their code currently can't accomplish that. [This](https://bugs.chromium.org/p/chromium/issues/detail?id=490015#c99) comment points out that this has broken other developers' tracking code as well, since it throws an exception now and doesn't even return false.

Thus, we can't send a beacon with `application/json` from chrome.

_(This seems to be a gap in the chromium network stack architecture, since hypothetically, the `fetch()` API should be able to use a `keepalive` flag for the exact same functionality, but this functionality [appears to be sidelined too](https://bugs.chromium.org/p/chromium/issues/detail?id=810466))_

Re: the collector - what I was saying is that the collector (correctly) only accepts POST requests with `content-type: "application/json"`, since the request actually is an `application/json`. Without changing that in the code, or putting our API Gateway hack in front of the collector that transforms `content-type`, there's no way to use the send-beacon-with-"plain/text"-content-type-from-chrome workaround. (Appreciate the info re: collector cors config, and that will probably come in handy for the next steps)

**Comment by mhadam** [comment]
I'm waiting on `2.10.1-rc2` to deploy, but it's updated to include a workaround for Chrome (it won't fix Beacon API, but I'll catch the exception and it should fallback to POST requests). As much as it's great that you're able to make it work with `plain/text`, I can't expect everyone else to set up their tracking pipeline to accommodate that. You're right though this seems like the Chrome team is (bizarrely) behind on this bug.

**Comment by pearsonhenri** [comment]
Just wanted to peep from the peanut gallery to say thanks for chasing this down. This was driving me bonkers when I tried to track down what was happening. Very arcane bug. Appreciate the quick patch @mhadam!
